### PR TITLE
[refactor] Remove dependencies on Callable::program in lang::CompiledGraph::run

### PR DESCRIPTION
**Pull request by PGZXB**
Issue: #7286
* https://github.com/taichi-dev/taichi/issues/7286: Part of https://github.com/taichi-dev/taichi/issues/7002

### CODE REVIEW

File: taichi/aot/graph_data.h [hunk 1]
```diff
@@ -161,8 +162,16 @@ struct TI_DLL_EXPORT CompiledGraph {
   RuntimeContext ctx_;
 
   void run(const std::unordered_map<std::string, IValue> &args) const;
+  void jit_run(const CompileConfig &compile_config,
```
Comment (ailzhang) [inline taichi/aot/graph_data.h:165]: @PGZXB do we want to expose a `jit_run` to external users? Say if a user call `jit_run` twice they will accidentally compile the kernels twice right? Maybe hiding the jit logic inside `run` will help?

**Comment by taichi-gardener** [comment]
Rebase failed, please resolve the conflicts manually.

**Pull request merged**